core/sequencer/sequencer.go

```diff
@@ -140,17 +144,19 @@ func (s *Sequencer) ListenForNewDomains(ctx context.Context, refresh time.Durati
 		case <-ticker.C:
 			domains, err := s.domains.List(ctx, false)
 			if err != nil {
-				return fmt.Errorf("admin.List(): %v", err)
+				return fmt.Errorf("domains.List(): %v", err)
 			}
 			for _, d := range domains {
 				knownDomains.Set(1, d.DomainID)
-				if _, ok := s.receivers[d.DomainID]; !ok {
-					glog.Infof("StartSigning domain: %v", d.DomainID)
-					r, err := s.NewReceiver(ctx, d)
-					if err != nil {
-						return err
-					}
-					s.receivers[d.DomainID] = r
+				if _, ok := s.controllers[d.DomainID]; !ok {
+					c := NewController(ctx, d, s.electionFactory)
+					s.controllers[d.DomainID] = c
+					go func() {
+						if err := c.RunWhileMaster(ctx, s.NewReceiver); err != nil {
+							glog.Errorf("RunWhileMaster: %v", err)
+						}
+						delete(s.controllers, d.DomainID)
```

Comment:
Deleting from `map` concurrently might be a problem?

Reply:
Yes, it's undefined behaviour. It will need synchronizing.

Reply:
Thanks for the catch